src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -393,6 +392,20 @@ categories :
       description : >-
         Striping will happen for message sizes beyond this threshold.
 
+    - name        : MPIR_CVAR_CH4_OFI_ENABLE_MULTIPLEXING
+      category    : CH4
+      type        : int
+      default     : 0
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_LOCAL
+      description : >-
+        Multiplexing means to use more than one NIC to send and receive messages above a certain
+        size.  If set to positive number, this feature will be turned on. If set to 0, this feature
+        will be turned off. If the number is -1, MPICH automatically determines whether to use
+        multiplexing depending on what is detected on the system (e.g., number of NICs available,
+        number of processes sharing the NICs).
+
```

Comment:
Multiplex means combining multiple messages/streams into a single channel, right? So I am not sure multiplex is the right word here. A more precise name maybe `MPIR_CVAR_CH4_OFI_MULTINIC_HASHING` (or something like that). 

At the user level, there is really not much difference with multi-nic vs multi-vni, both are just designating different separate endpoints. So maybe we could merge this multi-nic "multiplexing" feature as one of the `MPIDI_CH4_VCI_METHOD` and expand `MPIDI_get_vci` into `MPIDI_get_nic_vci`?

Reply:
I really appreciate you listening to me talk about this topic for one year before suggesting that I throw out all of the work and start over...

The naming comment is fine. You're probably right that multiplexing is an inaccurate way of describing it. I don't care what it's called.

As for the combination with VNI, I think mixing the notion of threads and multi-nic would be more confusing for the user. Users don't know what VNIs are since that's an MPICH-only concept. They do know what NICs and threads are since those are universal. When you offer to turn on features to make threads faster, it's clear what you're doing. That same is true with multiple NICs.

In the backend, these are implemented in the same way. We create additional endpoints/contexts when necessary and when we send/receive a message, we have a lookup function depending on which VNI and NIC pair you're using. The main differences (and the source of most of the code changes) come from the setup and address exchange of the NICs, not sending/receiving messages. We also have different optimizations for multiple NICs than we do for multiple VNIs. Trying to mix that code is unnecessary and adds much more complexity.

Reply:
Okay, I think it has some merits in separating the multi-nic from vci. The point of multi-nic is to increase the raw bandwidth, while the purpose of multi-vci is multi-threading and reduce the overhead. It is convenient to have nic hide inside vni so each (nic and vni) can take care of their respective purposes.

I like the "multiplex" name to be changed since we agree on that. The rest of the PR I believe is ready to merge.

Reply:
Ok. I've changed multiplexing to multi-nic hashing. To keep things consistent, I've also added a patch to change the name of the striping hint to multi-nic striping (helps differentiate from MPI I/O features and "namespaces" it with the hashing optimization. I left that as a separate commit since the striping code was at least partially already there.